### PR TITLE
[ADD] Country States for Switzerland

### PR DESCRIPTION
**Pull request by mncoon**
Little addon for Switzerland. Only States without cities.

### CODE REVIEW

File: l10n_ch_states/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from . import models
```
Comment (yvaucher) [inline l10n_ch_states/__init__.py]: Your editor is not putting a EOL before EOF. Please fix this to make travis happy. (and to respect POSIX standard)

**Comment by mncoon** [comment]
Thanks for the hint. Now all is well.

**Comment by yvaucher** [comment]
Please rename your README.rst with capital letters.

File: l10n_ch_states/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    'name': 'Switzerland Country States',
+    'category': 'Localisation',
+    'summary': '',
+    'version': '1.0',
+    'author': 'copado MEDIA UG, Odoo Community Association (OCA)',
+    'website': 'http://www.copado.de',
+    'license': 'AGPL-3',
+    'depends': ['base'],
+    'data': [],
+    'update_xml': ['data/res_country_states.xml'],
```
Comment (yvaucher) [inline l10n_ch_states/__openerp__.py]: `update_xml` is deprecated, use `data` instead

**Comment by yvaucher** [comment]
@nbessi https://github.com/OCA/l10n-switzerland/tree/8.0/l10n_ch_zip is already creating states right?

(file l10n_ch_better_zip.xml)

But here we have the translated names not only the codes

**Comment by jmvandel** [comment]
What happens if we install l10n_ch_zip after l10n_ch_states (and vice-versa)? Will we keep the state information on existing partners?
And what happens if both modules are left activated at the same time?

**Comment by yvaucher** [comment]
@jmvandel I think the way to go would be to extract state data from l10n_ch_zip and make it depend on this new module.

**Comment by mncoon** [comment]
@jmvandel Good question. From my point of view 10n_ch_states is an alternative for l10n_ch_zip for those who want only states without zip.

@yvaucher that is also a good idea. But than I should use the id's from l10n_ch_zip in l10n_ch_states.

If this is OK, I will do so.

**Comment by yvaucher** [comment]
to answer @jmvandel question having both modules with current code it would make duplicate states as there is no uniq constraint on res.country.state. And l10n_ch_zip wouldn't use those country states.

**Comment by open-net-sarl** [comment]
People WILL make mistakes and activate both modules or change their mind and activate one after using the other! 
We need to make sure that no data is lost in the process. ..

**Comment by nbessi** [comment]
I agree with @yvaucher on the concept as I do not want to have 2 addons having the same data.
But if we make a dependence between zip and state I will only accept to merge it if it has a solid migration script.

**Comment by mncoon** [comment]
@nbessi basic-work is done. Migration script will be done later.

**Comment by mncoon** [comment]
Does someon have a example or something, for a migration work. I never done this before and I would do it as best as possible. I'm gratefull to have some hints.

**Comment by yvaucher** [comment]
@copmn here are few examples of migration script:
https://github.com/OCA/l10n-switzerland/tree/8.0/l10n_ch_bank/migrations/8.0.9.0.0
https://github.com/OCA/l10n-canada/tree/8.0/res_partner_attributes_add_BN/migrations/8.0.1.0
https://github.com/OCA/purchase-workflow/tree/8.0/framework_agreement/migrations/8.0.2.0

**Comment by yvaucher** [comment]
@copmn You will need to bump  `l10n_ch_zip/__openerp__.py` version for migration script.

File: l10n_ch_states/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+{
+    'name': 'Switzerland Country States',
+    'category': 'Localisation',
+    'summary': '',
+    'description': '',
```
Comment (gurneyalex) [inline l10n_ch_states/__openerp__.py]: please remove empty description entry, and rename readme.rst to README.rst.

**Comment by mncoon** [comment]
Work is done. It was interesting to find out where the data were distributed in the database.
I hope the migration-script is as good as needed and handles all necessary data.

**Comment by nbessi** [comment]
@copmn you may want to rebase last commits :P. I'll try to have a look this friday

**Comment by mncoon** [comment]
@nbessi rebase is done but there is an error at travis on OCB-testing and I'm not sure what it means.

**Comment by yvaucher** [comment]
:+1:

**Comment by nbessi** [comment]
:+1:

**Comment by yvaucher** [comment]
@copmn Which error are you talking about? Maybe the travis or runbot was restarted in the mean time?

**Comment by mncoon** [comment]
@yvaucher on final testing there was an error on travis ocb that I can't explain, but few hours later everything was fine. So I think it has resolved itself.

**Comment by yvaucher** [comment]
@copmn ok thanks. Thank you for you contribution.

**Comment by open-net-sarl** [comment]
Hi! 

I think there is a bug with the migration script of l10n_ch_zip : 

| 2015-07-31 13:27:22,221 22015 INFO PS_test openerp.modules.module:
| module l10n_ch_zip: creating or updating database tables
| 2015-07-31 13:27:25,578 22015 INFO PS_test openerp.modules.loading:
| loading l10n_ch_zip/l10n_ch_better_zip.xml
| 2015-07-31 13:27:32,043 22015 INFO PS_test openerp.modules.migration:
| module l10n_ch_zip: Running migration [8.0.2.0.0>] post-migration
| 2015-07-31 13:27:32,043 22015 INFO PS_test upgrade: Migrating
| l10n_ch_zip from version 8.0.1.1.0
| 2015-07-31 13:27:32,044 22015 INFO PS_test upgrade: Updating
| res_better_zip from id 264 to id (368,)
| 2015-07-31 13:27:37,702 22015 INFO PS_test upgrade: Updating
| res_better_zip from id 165 to id (367,)
| 2015-07-31 13:27:38,366 22015 ERROR PS_test openerp.sql_db: bad
| query: DELETE FROM res_country_state WHERE id = 165;
| Traceback (most recent call last):
| File "/home/odoo/git/Odoo/odoo/openerp/sql_db.py", line 234, in
| execute
| res = self._obj.execute(query, params)
| IntegrityError: update or delete on table "res_country_state"
| violates foreign key constraint "res_partner_state_id_fkey" on table
| "res_partner"
| DETAIL: Key (id)=(165) is still referenced from table "res_partner".

| 2015-07-31 13:27:39,252 22015 CRITICAL PS_test
| openerp.service.server: Failed to initialize database `PS_test`.
| Traceback (most recent call last):
| File "/home/odoo/git/Odoo/odoo/openerp/service/server.py", line 929,
| in preload_registries
| registry = RegistryManager.new(dbname, update_module=update_module)
| File "/home/odoo/git/Odoo/odoo/openerp/modules/registry.py", line
| 370, in new
| openerp.modules.load_modules(registry._db, force_demo, status,
| update_module)
| File "/home/odoo/git/Odoo/odoo/openerp/modules/loading.py", line 351,
| in load_modules
| force, status, report, loaded_modules, update_module)
| File "/home/odoo/git/Odoo/odoo/openerp/modules/loading.py", line 255,
| in load_marked_modules
| loaded, processed = load_module_graph(cr, graph, progressdict,
| report=report, skip_modules=loaded_modules,
| perform_checks=perform_checks)
| File "/home/odoo/git/Odoo/odoo/openerp/modules/loading.py", line 183,
| in load_module_graph
| migrations.migrate_module(package, 'post')
| File "/home/odoo/git/Odoo/odoo/openerp/modules/migration.py", line
| 171, in migrate_module
| migrate(self.cr, pkg.installed_version)
| File
| "/home/odoo/git/OCA/l10n-switzerland/l10n_ch_zip/migrations/8.0.2.0.0/post-migration.py",
| line 66, in migrate
| "WHERE id = %s;", (old_state,))
| File "/home/odoo/git/Odoo/odoo/openerp/sql_db.py", line 158, in
| wrapper
| return f(self, _args, *_kwargs)
| File "/home/odoo/git/Odoo/odoo/openerp/sql_db.py", line 234, in
| execute
| res = self._obj.execute(query, params)
| IntegrityError: update or delete on table "res_country_state"
| violates foreign key constraint "res_partner_state_id_fkey" on table
| "res_partner"
| DETAIL: Key (id)=(165) is still referenced from table "res_partner".

The migration script should update all res_partners BEFORE suppressing States or update the names without changing the ID... 

With all my best regards, 

Jean-Marc Vandel 
CTO - Ing. info. dipl. EPFL 

Open Net Sàrl 
Rue de l'Industrie 59 
CH-1030 Bussigny 

www.open-net.ch 
+41 79 212.65.94 
jmv@open-net.ch

**Comment by mncoon** [comment]
@open-net-sarl Thank you for this hint. I totally forgotten this. Update is underway.

**Pull request merged**